src/oslogin_utils.cc

```diff
@@ -1235,4 +1250,128 @@ bool ContinueSession(bool alt, const string& email, const string& user_token, co
+  }
+
+  return true;
+}
+
+static bool CreateGoogleUserFile(string users_filename) {
+  std::ofstream users_file;
+  users_file.open(users_filename.c_str());
+
+  // OS Login directories are created by another product, guest-agent.
+  // We should be sure a file is opened for writing.
+  if (users_file.is_open()) {
+    // this is only for creating an empty file
+    users_file.close();
+
+    chown(users_filename.c_str(), 0, 0);
+    chmod(users_filename.c_str(), S_IRUSR | S_IWUSR | S_IRGRP);
+    return true;
+  }
+
+  return false;
+}
+
+static bool CreateGoogleSudoersFile(string sudoers_filename, const char *user_name) {
+  std::ofstream sudoers_file;
+  sudoers_file.open(sudoers_filename.c_str());
+
+  // OS Login directories are created by another product, guest-agent.
+  // We should be sure a file is opened for writing.
+  if (sudoers_file.is_open()) {
+    sudoers_file << user_name << " ALL=(ALL) NOPASSWD: ALL\n";
+    sudoers_file.close();
+
+    chown(sudoers_filename.c_str(), 0, 0);
+    chmod(sudoers_filename.c_str(), S_IRUSR | S_IRGRP);
+    return true;
+  }
+
+  return false;
+}
+
+bool AuthorizeUser(const char *user_name, struct AuthOptions opts, string *user_response) {
+  bool users_file_exists, sudoers_exists;
+  string email, users_filename, sudoers_filename;
+  struct stat sudoers_buffer, buffer;
+
+  sudoers_buffer = buffer = { 0 };
+  users_file_exists = sudoers_exists = false;
+
+  if (!ValidateUserName(user_name)) {
+    return false;
+  }
+
+  // Call MDS "users?username=" endpoint.
+  if (!MDSGetUser(user_name, opts.security_key, user_response)) {
+    return false;
+  }
+
+  if (!ParseJsonToEmail(*user_response, &email) || email.empty()) {
+    return false;
+  }
+
+  users_filename = kUsersDir;
+  users_filename.append(user_name);
+  users_file_exists = !stat(users_filename.c_str(), &buffer);
+
+  // Couldn't apply "login" policy for the loggin in user, log it and deny.
+  if (!ApplyPolicy(user_name, email, "login", opts)) {
+    SysLogErr("Could not grant access to organization user: %s.", user_name);
+    if (users_file_exists) {
+      remove(users_filename.c_str());
+    }
+    return false;
+  }
+
+  // If we can't create users file we can't grant access, log it and deny.
+  if (!users_file_exists && CreateGoogleUserFile(users_filename)) {
+    SysLogErr("Failed to create user's file.");
+    return false;
+  }
+
+  sudoers_filename = kSudoersDir;
+  sudoers_filename.append(user_name);
+  sudoers_exists = !stat(sudoers_filename.c_str(), &sudoers_buffer);
+
+  if (ApplyPolicy(user_name, email, "adminLogin", opts)) {
+    // Best effor creating sudoers file, if we fail log it and grant access.
+    if (!sudoers_exists && !CreateGoogleSudoersFile(sudoers_filename, user_name)) {
+      SysLogErr("Could not grant sudo permissions to organization user %s."
+                " Sudoers file %s is not writable.", user_name, sudoers_filename.c_str());
+    }
+  } else {
+    remove(sudoers_filename.c_str());
```

Comment:
are we deleting the whole file if the apply policy fails? is that what we want to do?

Reply:
I kept compatibility with the logic that was there before, now that you raised the concern I agree with you, we should be removing only the user's entry. I'll get it dealt with (not sure if in a followup PR or not).

Reply:
Just double checked, we have one sudoer file per user, so yes, it's safe to remove the whole file (it's equivalent to removing a line from a central database).